core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java

```diff
@@ -0,0 +1,405 @@
+                newScheduledThreadPool(10, threadsNamed("task-notification-%s")),
+                DataSize.of(32, DataSize.Unit.MEGABYTE),
+                1);
+
+        scaledWriterScheduler.schedule();
+
+        assertEquals(scaledWriterScheduler.schedule().getNewTasks().size(), 0);
+    }
+
+    @Test
+    public void testGetNewTaskCountWithUOverutilizedNonSkewedTasks()
+    {
+        TaskInfo taskInfo1 = buildTaskInfo(true, 12345L);
+        TaskInfo taskInfo2 = buildTaskInfo(true, 12345L);
+        TaskInfo taskInfo3 = buildTaskInfo(false, 12345L);
+
+        ScaledWriterScheduler scaledWriterScheduler = new ScaledWriterScheduler(
+                new TestingStageExecution(createFragment()),
+                () -> ImmutableList.of(taskInfo1, taskInfo2, taskInfo3),
+                () -> ImmutableList.of(taskInfo1.getTaskStatus(), taskInfo2.getTaskStatus(), taskInfo3.getTaskStatus()),
+                new UniformNodeSelectorFactory(
+                        new InMemoryNodeManager(NODE_1, NODE_2),
+                        new NodeSchedulerConfig().setIncludeCoordinator(true),
+                        new NodeTaskMap(new FinalizerService())).createNodeSelector(testSessionBuilder().build(), Optional.empty()),
+                newScheduledThreadPool(10, threadsNamed("task-notification-%s")),
+                DataSize.of(32, DataSize.Unit.MEGABYTE),
+                1);
+
+        scaledWriterScheduler.schedule();
+
+        assertEquals(scaledWriterScheduler.schedule().getNewTasks().size(), 1);
+    }
+
+    @Test
+    public void testGetNewTaskCountWithOverutilizedSkewedTaskAndUnderutilizedNonSkewedTasks()
+    {
+        TaskInfo taskInfo1 = buildTaskInfo(false, 12345L);
+        TaskInfo taskInfo2 = buildTaskInfo(false, 123456L);
+        TaskInfo taskInfo3 = buildTaskInfo(true, 1234567L);
+
+        ScaledWriterScheduler scaledWriterScheduler = new ScaledWriterScheduler(
+                new TestingStageExecution(createFragment()),
+                () -> ImmutableList.of(taskInfo1, taskInfo2, taskInfo3),
+                () -> ImmutableList.of(taskInfo1.getTaskStatus(), taskInfo2.getTaskStatus(), taskInfo3.getTaskStatus()),
+                new UniformNodeSelectorFactory(
+                        new InMemoryNodeManager(NODE_1, NODE_2),
+                        new NodeSchedulerConfig().setIncludeCoordinator(true),
+                        new NodeTaskMap(new FinalizerService())).createNodeSelector(testSessionBuilder().build(), Optional.empty()),
+                newScheduledThreadPool(10, threadsNamed("task-notification-%s")),
+                DataSize.of(32, DataSize.Unit.MEGABYTE),
+                1);
+
+        scaledWriterScheduler.schedule();
+
+        assertEquals(scaledWriterScheduler.schedule().getNewTasks().size(), 1);
+    }
+
+    @Test
+    public void testGetNewTaskCountWithUnderutilizedSkewedTaskAndOverutilizedNonSkewedTasks()
```

Comment:
What's the difference between this and the above test? I think this is not required

Reply:
Same as above.

Reply:
maybe even `testScheduleWithOverutilizedNonSkewedTasks`